### PR TITLE
Muparser -- Add as separate library

### PR DESCRIPTION
**Pull request by leamas**
Refactor the muparser code into a separate library, re-using the existing CMake setup to build it.

Builds successfully at least at circleci, havn't looked that much into it besides that.

### CODE REVIEW

**Comment by leamas** [comment]
fails on windows.Linkage errors indicates that Debug and Release libraries are mixed up. I cannot reproduce this error on my own windows box, so hard to tell what it's really  all about.

**Comment by leamas** [comment]
Found it, and it is an issue with the templates.

**Comment by leamas** [comment]
Installable on Flatpak. However, any attempt to click on the toolbar calculator icon causes a segfault (below). I don't think this is a build problem, more of a programming error in the plugin.

---
```
x00007ffff676a5b3 in _Unwind_Resume () from /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
(gdb) bt
#0  0x00007ffff676a5b3 in _Unwind_Resume () at /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
#1  0x00007ffff1c20b5c in __gnu_cxx::new_allocator<char>::~new_allocator() (this=<optimized out>, __in_chrg=<optimized out>) at /usr/include/c++/10.2.0/ext/new_allocator.h:89
#2  std::allocator<char>::~allocator() (this=<optimized out>, __in_chrg=<optimized out>) at /usr/include/c++/10.2.0/bits/allocator.h:162
#3  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Alloc_hider::~_Alloc_hider() (this=<optimized out>, __in_chrg=<optimized out>)
    at /usr/include/c++/10.2.0/bits/basic_string.h:150
#4  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() (this=<optimized out>, __in_chrg=<optimized out>)
    at /usr/include/c++/10.2.0/bits/basic_string.h:658
#5  Dlg::WxString2StdString[abi:cxx11](wxString) (this=<optimized out>, wxString_in=...) at /run/build/calculator/src/calculatorgui_impl.cpp:907
#6  0x00007ffff1c46ca4 in Dlg::Dlg(wxWindow*, calculator_pi*) (this=<optimized out>, parent=<optimized out>, ppi=0x555556c29430) at /run/build/calculator/src/calculatorgui_impl.cpp:390
#7  0x00007ffff1c2c06c in calculator_pi::OnToolbarToolCallback(int) (this=0x555556c29430, id=<optimized out>) at /run/build/calculator/src/calculator_pi.cpp:234
#8  0x00005555558c456f in MyFrame::OnToolLeftClick(wxCommandEvent&) ()
#9  0x00007ffff7176dee in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) () at /app/lib/libwx_baseu-3.0.so.0
#10 0x00007ffff7176eeb in wxEventHashTable::HandleEvent(wxEvent&, wxEvtHandler*) () at /app/lib/libwx_baseu-3.0.so.0
#11 0x00007ffff7177240 in wxEvtHandler::TryHereOnly(wxEvent&) () at /app/lib/libwx_baseu-3.0.so.0
#12 0x00007ffff71770a3 in wxEvtHandler::DoTryChain(wxEvent&) () at /app/lib/libwx_baseu-3.0.so.0
#13 0x00007ffff71773a1 in wxEvtHandler::ProcessEvent(wxEvent&) () at /app/lib/libwx_baseu-3.0.so.0
#14 0x00007ffff7177ea8 in wxEvtHandler::ProcessPendingEvents() () at /app/lib/libwx_baseu-3.0.so.0
#15 0x00007ffff7005c1f in wxAppConsoleBase::ProcessPendingEvents() () at /app/lib/libwx_baseu-3.0.so.0
#16 0x00007ffff74ca939 in wxApp::DoIdle() () at /app/lib/libwx_gtk3u_core-3.0.so.0
#17 0x00007ffff74caa67 in  () at /app/lib/libwx_gtk3u_core-3.0.so.0
#18 0x00007ffff530f75f in g_main_context_dispatch () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#19 0x00007ffff530fb08 in g_main_context_iterate.constprop () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#20 0x00007ffff530fe23 in g_main_loop_run () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#21 0x00007ffff577bebd in gtk_main () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#22 0x00007ffff74e4f35 in wxGUIEventLoop::DoRun() () at /app/lib/libwx_gtk3u_core-3.0.so.0
#23 0x00007ffff70469e1 in wxEventLoopBase::Run() () at /app/lib/libwx_baseu-3.0.so.0
#24 0x00007ffff70069e9 in wxAppConsoleBase::MainLoop() () at /app/lib/libwx_baseu-3.0.so.0
#25 0x00007ffff708ec8c in wxEntry(int&, wchar_t**) () at /app/lib/libwx_baseu-3.0.so.0
#26 0x00005555557cddfc in main ()
```

Looking at src/calculatorgui_impl.cpp:907  I see

```
mu::string_type Dlg::WxString2StdString(wxString wxString_in){
        std::string s = std::string(wxString_in.mb_str());
        std::wstring ws;
        UTF82WC(s, ws);
```

wxString's mb_str() is actually a known source of problems. This particular case might be easy to fix using wxString::ToStdString(), but a codebase with these kind of errors is somewhat scary.

EDIT: This function should probably be removed completely, replacing all invocations with `wxString::ToStdString()`which does the job without segfaults.

**Comment by leamas** [comment]
Following this path I find things like `WxString2StdString(_T("pi")` which is an insane way of writing `std::string("pi")`.  This code is in a really bad shape and probably needs an overall review.

**Comment by Rasbats** [comment]
My fork was only intended for Windows and some (all?) of this stuff may be my stuff. SaltyPaws made the original code so revisiting that may be useful for me.

**Comment by leamas** [comment]
Well, the basics of strings when using wxWidgets is that they Just Work (tm). There should be no need to make explicit conversions to/from wide strings except perhaps when calling windows API functions directly without going through wxWidgets. I can't see why this should happen here (?).

Overall, std::string is much more stable and should be preferred except when making wxWidgets calls. Conversions between wxString and std:.string are available both ways, notably `wxString::ToStdString`  and `wxString(const& std::string)`. 

In particular: Drop all useless _T() macros, and try to avoid using mb_str() which is kind of a problem child. It is possible to use it, but then you need to be aware of the ramifications. Ramifications which not are caught by the compiler.

I also doubt anything like `mu::string_type` is necessary or useful.

**Comment by leamas** [comment]
hm... quick look at the code seems to mean that it uses const char* to pass strings to the expression parser. The weird thing is that wxString..mb_str() will/might cause segfaults when used for this. The safe route to get a `const char*` from a wxString is `wxString::ToStdString().c_str()`

**Comment by Rasbats** [comment]
>fails on windows.Linkage errors indicates that Debug and Release libraries are mixed up

```
muparser.lib(muParserTokenReader.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in calculator_pi.obj [C:\project\opencpn\calculator_pi\build\calculator_pi.vcxproj] [C:\project\opencpn\calculator_pi\build\tarball-build.vcxproj]
muparser.lib(muParserTokenReader.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MD_DynamicRelease' in calculator_pi.obj [C:\project\opencpn\calculator_pi\build\calculator_pi.vcxproj] [C:\project\opencpn\calculator_pi\build\tarball-build.v
```

**Comment by Rasbats** [comment]
Fixed a local build with this:
[https://stackoverflow.com/questions/7668200/error-lnk2038-mismatch-detected-for-iterator-debug-level-value-0-doesnt](https://stackoverflow.com/questions/7668200/error-lnk2038-mismatch-detected-for-iterator-debug-level-value-0-doesnt)

**Comment by leamas** [comment]
This was fixed earlier, basically by changing `cmake ... --config xxx` to `cmake -DCMAKE_BUILD_TYPE=xxx`

EDIT: that is, in 987227a3dc325d6a7

**Comment by Rasbats** [comment]
Sorry ... I have hundreds of these bat files. Forgot about this fix.

**Comment by leamas** [comment]
> Sorry ...

Don't apologize. Really, don't.

**Comment by Rasbats** [comment]
Used  a cast - `string_type` instead of the function wxString2StandardString.
Now compiles for all OS.
Tested on Win, flatpak, Android and they work!
Looks good. Thanks for the lib fix.

**Comment by leamas** [comment]
Using a old-style C-cast might be pushing things quite hard when dealing with `wxString`. `std:.string` and `mu:stringtype`.

Looking at `string_type` there is first ` typedef MUP_STRING_TYPE string_type;` in  _muParserDef.h_. In the same file, later, abridged:

```
#if defined(_UNICODE)
        #define MUP_STRING_TYPE std::wstring
#else
        #define MUP_STRING_TYPE std::string
#endif
```

and, not least, CMakeLists.txt:
```
option(ENABLE_WIDE_CHAR "Enable wide character support" OFF)
...
if(ENABLE_WIDE_CHAR)
  target_compile_definitions(muparser PUBLIC _UNICODE)
endif()
```

I think this has been about this option ENABLE_WIDE_CHAR. Once upon a time this might have been set to ON for whatever reason causing `string_type` to be a `std::wstring` causing all sorts of trouble. However, the library is compiled with the default setting ENABLE_WIDE_CHAR = OFF which means that `string_type` just is an alias for `std::string`.

So, to repeat: Using C-style casts carries a lot of risks here, basically hiding things for the compiler. OTOH, since `string_type` actually just is `std::string` you could use  `wxString::toStdString()` instead which is a type-safe operation giving the compiler a descent chance to sort things out.

If you did a modern, C++ `dynamic_cast<>()` please just forget this, a dynamic cast is safe if accepted by the compiler. 

TL; DR; nice that the cast works, but it seems risky...

**Comment by leamas** [comment]
Sorry, this is basically just about how to phrase things. What you did was to replace `WxString2StdString(_T("pi")` with `string_type("pi")` etc.  This is not a cast, a cast would have been like `static_cast<string_type>("pi")` or even worse `(string_type)"pi"`, none of which I think should pass the compiler.

`string_type("pi")`  is a constructor; using that is perfectly fine and basically the same as `std::string("pi")`. In this context  I agree that using`string_type`is the better option.
